### PR TITLE
Add docs notification if GitHub oAuth is not setup

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
* If GitHub oauth is not set-up, show a notification with a link to a guide how to set-up GitHub identity provider in keycloak in Che docs page.
* Fix authentication mechanism (if Che user token is expired, retrieve it again), little refactoring.

### What issues does this PR fix or reference?

fixes https://github.com/eclipse/che/issues/18237

### How to test this PR?

1. Run Che with `quay.io/eclipse/che-server:nightly` image. It contains the [authentication fix](https://github.com/eclipse/che/pull/18815)
2. Create and open a workspace from dev-file:
```
apiVersion: 1.0.0
metadata:
  name: testnjz3q
components:
  - type: cheEditor
    reference: 'https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-969/che_theia_meta.yaml'
  - id: github/vscode-pull-request-github/latest
    type: chePlugin
projects:
  - name: environment-variable
    source:
      location: 'https://github.com/vinokurig/environment-variable.git'
      type: git
```
3. Try to login the GitHub PR plugin:
screenshot-serverkcgc53mh-che-dev-server-3010 192 168 99 254 nip io-2021 01 22-13_21_04
4. See the notification:
screenshot-serverkcgc53mh-che-dev-server-3010 192 168 99 254 nip io-2021 01 22-13_23_02

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

**Comment by sunix** [comment]
I suggest something like that:

`Could not start the Github Pull Request plugin. The setup for Github OAuth provider is not complete.`

And only single button: `Open the setup instructions page`

**Comment by vinokurig** [comment]
@sunix 
> I suggest something like that:
>
>`Could not start the Github Pull Request plugin. The setup for Github OAuth provider is not complete.`

It is not specific to the `Github Pull Requests plugin` so I would keep only:  `The setup for Github OAuth provider is not complete.`
>And only single button: `Open the setup instructions page`

Notification buttons are not very flexible in theia, so it will look like:
screenshot-che-eclipse-che 192 168 99 254 nip io-2021 01 26-15_42_00

**Comment by vinokurig** [comment]
@benoitf @l0rd Could you please take a look?

File: extensions/eclipse-che-theia-remote-api/src/browser/oauth-utils.ts [hunk 1]
```diff
@@ -57,29 +59,41 @@ export class OauthUtils {
...
-      const timer = setTimeout(() => {
-        this.messageService.warn(
-          'Authentication is taking too long, the oauth pop-up may be blocked by your browser, ' +
-            'if so, allow popup windows for the current url and restart the workspace'
+    const updateToken: () => Promise<string | undefined> = async () => {
+      if (this.machineToken && this.machineToken.length > 0) {
+        const timer = setTimeout(() => {
+          this.messageService.warn(
+            'Authentication is taking too long, the oauth pop-up may be blocked by your browser, ' +
+              'if so, allow popup windows for the current url and restart the workspace'
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-remote-api/src/browser/oauth-utils.ts]: fixed

File: plugins/github-auth-plugin/src/github-auth-plugin.ts [hunk 2]
```diff
@@ -23,6 +23,21 @@ export async function start(context: theia.PluginContext): Promise<void> {
...
+        if (
+          (await theia.window.showWarningMessage(
+            'GitHub OAuth provider is not setup! Would you like to open the documentation page?',
+            'Yes',
+            'No'
+          )) === 'Yes'
+        ) {
+          theia.commands.executeCommand(
+            'theia.open',
+            'https://www.eclipse.org/che/docs/che-7/administration-guide/configuring-authorization/#configuring-github-oauth_che'
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply plugins/github-auth-plugin/src/github-auth-plugin.ts:35]: fixed

File: extensions/eclipse-che-theia-remote-api/src/browser/oauth-utils.ts [hunk 3]
```diff
@@ -57,29 +59,41 @@ export class OauthUtils {
...
-    } else if (this.machineToken && this.machineToken.length > 0) {
-      const timer = setTimeout(() => {
-        this.messageService.warn(
-          'Authentication is taking too long, the oauth pop-up may be blocked by your browser, ' +
-            'if so, allow popup windows for the current url and restart the workspace'
+    const updateToken: () => Promise<string | undefined> = async () => {
+      if (this.machineToken && this.machineToken.length > 0) {
+        const timer = setTimeout(() => {
+          this.messageService.warn(
+            'Authentication is taking too long, the oauth pop-up may be blocked by your browser, ' +
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-remote-api/src/browser/oauth-utils.ts]: fixed

File: plugins/github-auth-plugin/src/github-auth-plugin.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinokurig) [reply plugins/github-auth-plugin/src/github-auth-plugin.ts:35]: I think the eclipse docs page is relevant for downstream as well.

**Review by azatsarynnyy — COMMENTED** [review]
hm.. for some reasons, I see a bit different `PULL REQUEST` view
image

**Comment by vinokurig** [comment]
@azatsarynnyy Looks like there is no git project in your workspace. It's my mistake, I haven't included it to the example `devfile`, now it is updated :).

**Review by azatsarynnyy — APPROVED** [review]
The flow works as described. Thanks!

**Comment by vinokurig** [comment]
[crw-ci-test]

**Comment by azatsarynnyy** [comment]
fyi: HappyPath running depends on the fix https://github.com/eclipse/che/pull/18918

**Comment by dmytro-ndp** [comment]
[crw-ci-test]

**Comment by sunix** [comment]
> @sunix
> 
> > I suggest something like that:
> > `Could not start the Github Pull Request plugin. The setup for Github OAuth provider is not complete.`
> 
> It is not specific to the `Github Pull Requests plugin` so I would keep only: `The setup for Github OAuth provider is not complete.`
> 
> > And only single button: `Open the setup instructions page`
> 
> Notification buttons are not very flexible in theia, so it will look like:
> screenshot-che-eclipse-che 192 168 99 254 nip io-2021 01 26-15_42_00

in the message you have to tell
1. the problem
2. the cause
3. the solution

OK so something like

`Che could not authenticate to your Github account. The setup for Github OAuth provider is not complete.`

For the button, avoid `...`
So this should be enough: `Setup instructions`

**Comment by vinokurig** [comment]
@sunix
I like your message, applied it, thanks

**Pull request merged**